### PR TITLE
Add database functionality

### PR DESCRIPTION
**Pull request by TomDijkema**
Adds database functionality to the application.

Add:
- Function for inserting records into the countries or organisations table
- Function for requesting countries or organisations and reformatting data

Modify:
- Graph functions now rely on the database functionality instead of CSV
- CSV functions are now suppressed but still remain in the repository

### CODE REVIEW

**Comment by TomDijkema** [comment]
I will look at the complexity of functions, figure out a way to simplify them

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+# Copy the dependencies file to the working directory
+COPY requirements.txt .
+
+# Install dependencies
+RUN pip install -r requirements.txt
+
+# Copy the content of the local src directory to the working directory
+COPY . .
+
+# Adding file permission for csv files
```
Comment (samleeflang) [inline Dockerfile]: Is this still necessary?

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,29 @@
+# Set base image (host OS)
+FROM python:3.10-slim
+
+# Set the working directory in the container
+WORKDIR /code
+
+# Set csv_files directory as mounting volume
```
Comment (samleeflang) [inline Dockerfile]: Maybe just remove them if it is no longer necessary

File: query_database.py [hunk 3]
```diff
@@ -0,0 +1,405 @@
+import psycopg2
```
Comment (samleeflang) [inline query_database.py]: https://www.sqlalchemy.org/ probably makes a couple of things easier.
It is also database agnostic so if we want to switch to something is it is a bit more generic

File: kubernetes.yaml [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    matchLabels:
+      app: network-status-overview
+  schedule: "0 0 2 * *"
+  jobTemplate:
+    spec:
+      template:
+        spec:
+          containers:
+            - name: network-status-overview
+              image: network-status-overview
```
Comment (samleeflang) [inline kubernetes.yaml:18]: I think image should be the full path to the AWS image repo. Image is probably not pushed to that yet.

File: main2.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
+import logging
+
+
+# Import database logic
+# import query_database
```
Comment (samleeflang) [inline main2.py]: Remove unused import

File: main2.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+import GBIF_functions
+# GeoCASe API functionality
+import GeoCASe_functions
+import query_database
+
+logging.basicConfig(format='%(asctime)s - %(message)s', level=logging.INFO)
+
+
+def process_countries():
+    logging.info('\nReceiving countries datasets data from GBIF...')
```
Comment (samleeflang) [inline main2.py]: The '\n' should be necessary to enforce a new line. Are you sure this is needed now you use the logger?

File: package-lock.json [hunk 7]
```diff
@@ -0,0 +1,11 @@
+{
```
Comment (samleeflang) [inline package-lock.json:1]: What does this file do?

File: main2.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    logging.info('\nReceiving organisations datasets data from GBIF...')
+    gbif_organisations_data = GBIF_functions.gather_institutions()
+
+    logging.info('\nReceiving organisations data from GeoCASe...')
+    geocase_data = GeoCASe_functions.gather_publishers()
+
+    query_database.insert_organisations_data(gbif_organisations_data, geocase_data)
+
+
+process_countries()
```
Comment (samleeflang) [inline main2.py]: Isn't a main function necessary here. If this becomes the primary main.py it is probably needed to start the scripts?

File: query_database.py [hunk 9]
```diff
@@ -0,0 +1,405 @@
...
+            }),
+            'issues_flags': json.dumps(gbif_issues_flags['countries'][country])
+        }
+
+        # Create database connection
+        params = config()
+        conn = psycopg2.connect(**params)
+        cur = conn.cursor()
+
+        # Check if entity already exists (for same month)
```
Comment (samleeflang) [inline query_database.py]: This is also possible with an upsert query:
https://www.postgresqltutorial.com/postgresql-tutorial/postgresql-upsert/

File: query_database.py [hunk 10]
```diff
@@ -0,0 +1,405 @@
...
+        row = cur.fetchone()
+
+        if row:
+            # Update existing record
+            query = update_countries(country_entity, row)
+        else:
+            # Insert new record
+            query = insert_countries(country_entity)
+
+        # Execute query
```
Comment (samleeflang) [inline query_database.py:104]: See if you can use the try-with-resources functionality then you don't need to explicitly commit and close the connection.
For sql alchemy:
https://docs.sqlalchemy.org/en/14/core/connections.html

**Review by samleeflang — COMMENTED** [review]
Nice that the database works and this looks good.
Check if you can remove the csv references some more so we don't leave unused code in there.
For db management switching to sqlalchemy would probably be good

File: query_database.py (hunk 10 above)
Comment (TomDijkema) [reply query_database.py:104]: Replaced plain Psycopg2 database queries with SQL Alchemy queries that use the 'with' method

File: query_database.py (hunk 9 above)
Comment (TomDijkema) [reply query_database.py]: Combined separate insert and update functions (queries) to one upset query for both ways

File: Dockerfile (hunk 1 above)
Comment (TomDijkema) [reply Dockerfile]: Removed CSV references in Docker file, however, Docker file does not properly start python script, could you please have a look at the changes of the Dockerfile in the new commit?

File: Dockerfile (hunk 2 above)
Comment (TomDijkema) [reply Dockerfile]: Removed CSV references in Docker file, however, Docker file does not properly start python script, could you please have a look at the changes of the Dockerfile in the new commit?

File: query_database.py (hunk 3 above)
Comment (TomDijkema) [reply query_database.py]: Replaced core psycopg2 usage with SQLAlcehmy, psycopg2 is still being used as the database driver for Postgresql and is now included in the Dockerfile which needed the special package libpg-dev and a C compiler (build-essential)

File: kubernetes.yaml (hunk 4 above)
Comment (TomDijkema) [reply kubernetes.yaml:18]: Image has not been pushed yet indeed, I shall insert a reference in the file for remembrance, probably good to first have a look at the Dockerfile before uploading the image to AWS

File: main2.py (hunk 5 above)
Comment (TomDijkema) [reply main2.py]: Done

File: main2.py (hunk 6 above)
Comment (TomDijkema) [reply main2.py]: Correct assumption, removed the '\n' elements

File: package-lock.json (hunk 7 above)
Comment (TomDijkema) [reply package-lock.json:1]: The ifle is important for the fixed range of versioning packages and is automatically generated by npm: 
https://nodejs.dev/learn/the-package-lock-json-file

File: main2.py (hunk 8 above)
Comment (TomDijkema) [reply main2.py]: Added main function, will be optimized in upcoming version when application will be split up into two smaller components

File: package-lock.json (hunk 7 above)
Comment (samleeflang) [reply package-lock.json:1]: Ok thanks.

**Review by samleeflang — APPROVED** [review]

**Pull request merged**